### PR TITLE
add Assertion, AssertionBase and Decode

### PR DESCRIPTION
**Pull request by pedronis**
so this one introduces the Assertion interface with enough to access the basic bits of their representation, AssertionBase that can be used/wrapped and can hold those bits concretely, and Decode that takes a serialized assertion and builds an Assertion of the right concrete type driven by the typeRegistry table that for now can list builders taking AssertionBase and making specific assertion structs but will grow more bits to drive more processing later.

Notice that for now I haven't introduced any concrete assertion types except a test-only one (TestOnly) in export_test.go .

The expected serialization format is based on the original "spec" but with some names adapted also taking into account the Design doc;  thing I dislike about it as it is atm:

'body-length' polluting the namespace of actual assertion fields, I don't know, it's mostly there for sanity  checking and later I think to allow having bundles of serialized assertions with just extra separators between them.

My current thinking is that we could maybe have some kind of prefix for fields like that so they cannot clash with actual/normal assertion fields... vaguely thinking of some macro assemblers and similar I'm thinking just using "." so:

.body-length: 6 

too obscure, cute?  I think we may be ok though with body-length because it sounds much clearly about the surrounding document itself than something plain generic as 'size' was.

### CODE REVIEW

**Comment by chipaca** [comment]
Overall, nice branch. :+1:

About metadata fields, maybe support comments, and use the widely understood -*- key: value; key: value -*- format for metadata in a comment?

**Comment by pedronis** [comment]
@chipaca well these things are not quite meant to be human produced, not sure I want to add the whole concept of comments there, and hide integrity bits in them?

any opinion on moving signature-type to prefix the signature blob itself?

**Comment by chipaca** [comment]
ah, i hadn't realized it was interity-related. Yeah, no.

signature-type prefix seems to be the standard, I think you'd have to come up with a reason why *not* to do it at this point.

**Comment by pedronis** [comment]
thanks @chipaca, ok, I'll rework signature-type to be a prefix, it seems it will much increase the consistency with other things later

File: asserts/asserts.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+// Understood assertions
+const (
+// IdentityType AssertionType = ...
+// AccountKeyType AssertionType = "account-key"
+// ...
+)
+
+// Assertion signature format types
+const (
+	OpenPGPSig = "openpgp"
```
Comment (niemeyer) [inline asserts/asserts.go]: Can we please drop this until we have a reason to differentiate?  I expect this to be quite far in the future, so it seems best to keep the design details to that future moment.

Comment (pedronis) [inline asserts/asserts.go]: @niemeyer I'm bit uncomfortable having just the blob and knowing only implicitly it's pgp, also from tyler discussion and the fact we haven't talked with agl yet it's still unclear whether we can really go all the way with pgp.

Comment (pedronis) [inline asserts/asserts.go]: @niemeyer we can keep the prefix though on the wire, but make it only a problem of checking and signing, just means later error catching maybe, so the signature of Signature gets back to just content, signature

Comment (niemeyer) [inline asserts/asserts.go]: Sorry, I now realize my comment was far from clear. I was suggesting dropping the exported constant, specifically, while it's the only possible constant and irrelevant for consumers of that API. The signature prefix looks fine.

File: asserts/asserts.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+	AuthorityID() string
+
+	// Header retrieves the header with name
+	Header(name string) string
+
+	// the body of this assertion
+	Body() []byte
+
+	// Signature returns the signed content and its signature already split and decoded
+	Signature() (content []byte, sigtype string, signature []byte)
```
Comment (niemeyer) [inline asserts/asserts.go]: Can we come back to the original design of just content and signature? It should be fine for the signature prefix to be parsed by whoever has to understand it, I think.

Comment (niemeyer) [inline asserts/asserts.go]: Okay, I see the problem you were trying to solve now. To do this we need to not decode the signature. This should be fine, I think. It also means the Signature method will return the payload almost untouched, which sounds like an interesting property.

File: asserts/asserts.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+	}
+	headers := make(map[string]string)
+	for _, entry := range strings.Split(string(head), "\n") {
+		nameValueSplit := strings.Index(entry, ": ")
+		if nameValueSplit == -1 {
+			return nil, fmt.Errorf("header entry missing name value ': ' separation: %q", entry)
+		}
+		name := entry[:nameValueSplit]
+		if !headerNameSanity.MatchString(name) {
+			return nil, fmt.Errorf("invalid header name: %v", name)
```
Comment (niemeyer) [inline asserts/asserts.go]: s/%v/%q/, please, so we can see any awkward content in there.

File: asserts/asserts.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+//      SIG-TYPE " " BASE64-SIG-PACKET
+//
+//    SIG-TYPE is the type/format expected for the base64
+//    encoded signature value, for now only "openpgp" is supported
+//
+// An header line looks like:
+//
+//   NAME ": " VALUE
+//
+// for sanity NAME is expected to match headerNameSanity regexp.
```
Comment (niemeyer) [inline asserts/asserts.go]: That variable is not exported, so we shouldn't talk about it in the public API. It also feels like we probably don't need to document this. The field names will come out of specifications we write, so the internal validation is just making us honest about our own conventions.

File: asserts/asserts.go (hunk 1 above)
Comment (pedronis) [inline asserts/asserts.go]: ah, yes, that I can do

File: asserts/asserts.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+//
+//    HEADER is a set of header lines separated by "\n"
+//    BODY can be arbitrary,
+//    SIGNATURE is composed like:
+//      SIG-TYPE " " BASE64-SIG-PACKET
+//
+//    SIG-TYPE is the type/format expected for the base64
+//    encoded signature value, for now only "openpgp" is supported
+//
+// An header line looks like:
```
Comment (niemeyer) [inline asserts/asserts.go]: s/An/A/

File: asserts/asserts.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+//
+//   NAME ": " VALUE
+//
+// for sanity NAME is expected to match headerNameSanity regexp.
+//
+// The following headers are mandatory:
+//
+//   type
+//   authority-id (the signer id)
+//   revision (a positive int)
```
Comment (niemeyer) [inline asserts/asserts.go]: It's not quite clear this is mandatory yet. It's okay to keep it here for now, I suppose.

File: asserts/asserts.go (hunk 2 above)
Comment (pedronis) [inline asserts/asserts.go]: ok, I'm fine with doing that. just means some errors will be a bit later but less upfront parsing complexity... really only checking/signing will have to care

File: asserts/asserts.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+//   body-length (int expected to be equal to the length of BODY)
```
Comment (niemeyer) [inline asserts/asserts.go]: If we make this default to zero, we might omit it when there's no body, which feels a bit nicer than stating body properties for something without one.

File: asserts/asserts.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+//
+func Decode(serializedAssertion []byte) (Assertion, error) {
```
Comment (niemeyer) [inline asserts/asserts.go]: Given this is security-sensitive, I think I'd be slightly more comfortable making a copy of the data here. It means further mutations of the underlying array cannot affect the generated assertion.

File: asserts/asserts.go (hunk 4 above)
Comment (pedronis) [inline asserts/asserts.go]: yes, didn't think on the spot it's awkward to reference that variable

File: asserts/asserts.go (hunk 7 above)
Comment (pedronis) [inline asserts/asserts.go]: sounds reasonable

File: asserts/asserts.go (hunk 8 above)
Comment (pedronis) [inline asserts/asserts.go]: you mean copy the array upfront before doing the parsing/splitting?

File: asserts/asserts.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+	headers, err := parseHeaders(head)
+	if err != nil {
+		return nil, fmt.Errorf("parsing assertion headers: %v", err)
+	}
+
+	if len(signature) == 0 {
+		return nil, fmt.Errorf("empty assertion signature")
+	}
+
+	sigtypeSigpacketSplit := bytes.IndexByte(signature, ' ')
```
Comment (niemeyer) [inline asserts/asserts.go]: Given comments above, all the decoding from here on would be moved into the signature-checking logic.

Comment (pedronis) [inline asserts/asserts.go]: @niemeyer exactly, much less code (and error cases) for the moment actually

File: asserts/asserts.go [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+		return nil, fmt.Errorf("could not base64 decode the assertion signature packet")
+	}
+	sigpacket = decodedSigpacket[:n]
+
+	return buildAssertion(headers, body, content, sigtype, sigpacket)
+}
+
+func buildAssertion(headers map[string]string, body, content []byte, sigtype string, sigpacket []byte) (Assertion, error) {
+
+	checkInteger := func(name string) (int, error) {
```
Comment (niemeyer) [inline asserts/asserts.go]: I suggest moving this into a top-level function and taking header as a parameter instead. Makes the logic nicely isolated and reusable.

File: asserts/asserts.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+
+	length, err := checkInteger("body-length")
+	if err != nil {
+		return nil, err
+	}
+	if length != len(body) {
+		return nil, fmt.Errorf("assertion body length and declared body-length don't match: %v != %v", len(body), length)
+	}
+
+	checkMandatory := func(name string) (string, error) {
```
Comment (niemeyer) [inline asserts/asserts.go]: Same as above. This way we'll take 20 lines out of a pretty critical function in the system, making the important logic more readable.

File: asserts/asserts_test.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+	"github.com/ubuntu-core/snappy/asserts"
+)
+
+type AssertsSuite struct{}
+
+var _ = Suite(&AssertsSuite{})
+
+func (as *AssertsSuite) TestDecodeEmptyBody(c *C) {
+	encoded := "type: test-only\n" +
+		"revision: 0\n" +
```
Comment (niemeyer) [inline asserts/asserts_test.go]: I'd standardize having the revision after the primary key, so:

```
type (what is this)
authority-id (who said so)
primary-key (about what)
revision (how recently)
cruft (whatever)
```

It also makes me wonder again about the fact revision may not be always there. Might be worth going over the list of known assertions in the design doc. I'm almost sure we talked about one case where there was no revision. Can't remember the reason now, but looking at the assertion I probably will.

File: asserts/asserts_test.go [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+		"authority-id: auth-id1\n" +
+		"body-length: 0" +
+		"\n\n" +
+		"openpgp c2ln"
+	a, err := asserts.Decode([]byte(encoded))
+	c.Assert(err, IsNil)
+	c.Check(a.Type(), Equals, asserts.AssertionType("test-only"))
+	_, ok := a.(*asserts.TestOnly)
+	c.Check(ok, Equals, true)
+	c.Check(a.Revision(), Equals, 0)
```
Comment (niemeyer) [inline asserts/asserts_test.go]: As a side note, if we decide to make revision optional, I'd be fine with having Revision returning 0 still in those cases.

**Comment by chipaca** [comment]
:+1: once again :-)

File: asserts/asserts_test.go (hunk 13 above)
Comment (pedronis) [inline asserts/asserts_test.go]: that's what is done now in fact

File: asserts/asserts_test.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+	c.Check(a.AuthorityID(), Equals, "auth-id1")
+	cont, signature := a.Signature()
+	c.Check(signature, DeepEquals, []byte("openpgp c2ln"))
+	c.Check(cont, DeepEquals, []byte(content))
+}
+
+func (as *AssertsSuite) TestDecodeNoSignatureSplit(c *C) {
+	for _, encoded := range []string{
+		"",
+		"foo",
```
Comment (niemeyer) [inline asserts/asserts_test.go]: That's sooo much better as `[]string{"", "foo"}`. :-)

File: asserts/asserts.go [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+// Understood assertions
+const (
+// IdentityType AssertionType = ...
+// AccountKeyType AssertionType = "account-key"
+// ...
+)
+
+// Assertion represents an assertion through its general elements.
+type Assertion interface {
+	// the type of this assertion
```
Comment (niemeyer) [inline asserts/asserts.go]: These should probably be done in the usual wording style:
```
// Type returns the assertion type.
```

**Comment by niemeyer** [comment]
Looks great. I've made a couple of very trivial comments, but it's okay to apply these in the follow up.

**Pull request merged**